File: main/livekit-server/src/services/mem0_service.py
"""
Mem0 Service - Long-term memory management for Cheeko agents

This service handles:
- Storing conversation transcripts to Mem0 for fact extraction
- Retrieving memories for prompt injection
- Knowledge graph with entities and relations
"""

import os
import logging
import asyncio
from typing import List, Dict, Any, Optional

logger = logging.getLogger("mem0_service")


class Mem0Service:
    """
    Service for interacting with Mem0 API for persistent memory.
    Handles conversation storage and fact extraction with knowledge graph support.
    """

    def __init__(self):
        self.api_key = os.getenv("MEM0_API_KEY")
        self.client = None

        logger.info(f"[MEM0] Initializing Mem0Service... API key present: {bool(self.api_key)}")

        if not self.api_key:
            logger.warning("[MEM0] MEM0_API_KEY not set. Memory features disabled.")
        else:
            try:
                from mem0 import MemoryClient
                self.client = MemoryClient(api_key=self.api_key)
                logger.info("[MEM0] Service initialized successfully with graph support")
            except ImportError as e:
                logger.error(f"[MEM0] mem0 package not installed. Run: pip install mem0ai. Error: {e}")
            except Exception as e:
                logger.error(f"[MEM0] Failed to initialize: {e}")

    def _normalize_user_id(self, user_id: str) -> str:
        """Normalize MAC address to consistent format (lowercase, with colons)"""
        # Keep colons, just lowercase
        return user_id.lower()

    async def add_conversation(
        self,
        user_id: str,
        messages: List[Dict[str, Any]],
        session_id: str = None
    ) -> bool:
        """
        Send conversation transcript to Mem0 for fact extraction.
        This is called at session end (fire-and-forget).

        Args:
            user_id: Device MAC address
            messages: List of {chatType: 1|2, content: str, timestamp: int}
            session_id: Optional session identifier for tracking

        Returns:
            bool: True if successful, False otherwise
        """
        logger.info(f"[MEM0] add_conversation called - user: {user_id}, messages: {len(messages) if messages else 0}, client_ready: {self.client is not None}")

        if not self.client:
            logger.warning("[MEM0] Client not initialized, skipping add_conversation")
            return False

        if not messages:
            logger.debug("[MEM0] No messages to add")
            return True

        try:
            clean_user_id = self._normalize_user_id(user_id)

            # Format messages for Mem0 API
            # chatType: 1 = user, 2 = agent
            formatted_messages = []
            for msg in messages:
                content = msg.get("content", "").strip()
                if content:
                    role = "user" if msg.get("chatType") == 1 else "assistant"
                    formatted_messages.append({
                        "role": role,
                        "content": content
                    })

            if not formatted_messages:
                logger.debug("[MEM0] No valid messages after filtering")
                return True

            # Add to Mem0 with graph extraction enabled
            # Mem0's sync client wrapped for async execution
            loop = asyncio.get_event_loop()
            await loop.run_in_executor(
                None,
                lambda: self.client.add(
                    formatted_messages,
                    user_id=clean_user_id,
                    enable_graph=True,  # Enable knowledge graph extraction
                    metadata={"session_id": session_id} if session_id else None
                )
            )

            logger.info(f"[MEM0] Added {len(formatted_messages)} messages for user {clean_user_id}")
            return True

        except Exception as e:
            logger.error(f"[MEM0] Failed to add conversation: {e}")
            return False

    async def get_memories(
        self,
        user_id: str,
        limit: int = 20
    ) -> Dict[str, Any]:
        """
        Retrieve memories for a user using semantic search with graph support.

        Args:
            user_id: Device MAC address
            limit: Maximum number of memories to retrieve (~500 tokens)

        Returns:
            Dict with: memories (list of strings), relations (list), entities (list)
        """
        if not self.client:
            return {"memories": [], "relations": [], "entities": []}

        try:
            clean_user_id = self._normalize_user_id(user_id)

            # Use semantic search with graph enabled
            loop = asyncio.get_event_loop()
            results = await loop.run_in_executor(
                None,
                lambda: self.client.search(
                    "What is known about this person, their family, pets, interests, skills, routines, and feelings?",
                    user_id=clean_user_id,
                    limit=limit,
                    enable_graph=True
                )
            )

            if results:
                # Extract memories
                memories = []
                entities = []

                for r in results.get("results", []):
                    if r.get("memory"):
                        memories.append(r["memory"])
                    # Extract entities from each result
                    for e in r.get("entities", []):
                        if e.get("name"):
                            entities.append({
                                "name": e["name"],
                                "type": e.get("type", "unknown")
                            })

                relations = results.get("relations", [])

                logger.info(f"[MEM0] Retrieved {len(memories)} memories, {len(relations)} relations for {clean_user_id}")

                return {
                    "memories": memories,
                    "relations": relations,
                    "entities": entities
                }

            return {"memories": [], "relations": [], "entities": []}

        except Exception as e:
            logger.error(f"[MEM0] Failed to get memories: {e}")
            return {"memories": [], "relations": [], "entities": []}

    async def add_fact(self, user_id: str, fact: str) -> bool:
        """
        Add a single fact to memory.

        Args:
            user_id: Device MAC address
            fact: The fact to store (e.g., "Loves dinosaurs")

        Returns:
            bool: True if successful
        """
        if not fact:
            return False

        return await self.add_conversation(
            user_id,
            [{"chatType": 1, "content": fact}]
        )

    async def search_relevant_memories(
        self,
        user_id: str,
        query: str,
        limit: int = 5
    ) -> List[str]:
        """
        Search memories relevant to the current conversation context.
        Called in real-time before each LLM response to find contextually relevant memories.

        Args:
            user_id: Device MAC address
            query: The current user message/query (e.g., "tell me a dog story")
            limit: Maximum number of relevant memories to retrieve

        Returns:
            List of memory strings relevant to the query
        """
        if not self.client or not query:
            return []

        try:
            clean_user_id = self._normalize_user_id(user_id)

            logger.info(f"[MEM0] Searching relevant memories for query: '{query[:50]}...'")

            # Use the user's actual query for semantic search
            loop = asyncio.get_event_loop()
            results = await loop.run_in_executor(
                None,
                lambda: self.client.search(
                    query,  # Use actual user message for semantic relevance
                    user_id=clean_user_id,
                    limit=limit,
                    enable_graph=True
                )
            )

            if results:
                memories = []
                for r in results.get("results", []):
                    if r.get("memory"):
                        memories.append(r["memory"])

                if memories:
                    logger.info(f"[MEM0] Found {len(memories)} relevant memories for query")
                return memories

            return []

        except Exception as e:
            logger.error(f"[MEM0] Failed to search relevant memories: {e}")
            return []

    def format_relevant_memories_for_injection(
        self,
        query: str,
        memories: List[str]
    ) -> str:
        """
        Format relevant memories as a context injection for the LLM.

        Args:
            query: The user's current query
            memories: List of relevant memory strings

        Returns:
            Formatted string to inject into the conversation context
        """
        if not memories:
            return ""

        formatted = f"""
🧠 RELEVANT MEMORIES FOR THIS REQUEST:
The child just said: "{query}"

YOU MUST USE these memories in your response:
"""
        for memory in memories:
            formatted += f"- {memory}\n"

        formatted += """
⚠️ CRITICAL: If any memory above is relevant to what the child asked, you MUST incorporate it!
Example: If they asked about a dog and you know their dog's name → USE that name!
"""
        return formatted

    def is_ready(self) -> bool:
        """Check if the Mem0 client is initialized and ready"""
        return self.client is not None


# Singleton instance for use across workers
mem0_service = Mem0Service()

File: main/manager-api-node/src/services/integrations/mem0.service.js
/**
 * Mem0 Memory Service
 *
 * Handles memory/personalization operations using the Mem0 API.
 * Used for storing and retrieving user preferences, facts, and conversation history.
 * Supports knowledge graph with entities and relations.
 */

const logger = require('../../utils/logger');

// Environment variables
const MEM0_API_KEY = process.env.MEM0_API_KEY;
// Normalize API URL - remove trailing slash and /memories/ suffix if present
let MEM0_API_URL = process.env.MEM0_API_URL || 'https://api.mem0.ai/v1';
MEM0_API_URL = MEM0_API_URL.replace(/\/+$/, '').replace(/\/memories\/?$/, '');

// Configuration
const MEM0_MEMORY_LIMIT = parseInt(process.env.MEM0_MEMORY_LIMIT, 10) || 20;
const MEM0_TIMEOUT_MS = parseInt(process.env.MEM0_TIMEOUT_MS, 10) || 5000;

/**
 * HTTP client for Mem0 API
 * Uses native fetch API available in Node.js 18+
 */
let httpClient = null;

/**
 * Initialize HTTP client for Mem0 API
 * @returns {Object|null} HTTP client configuration or null if not configured
 */
const getClient = () => {
  if (httpClient) {
    return httpClient;
  }

  if (!MEM0_API_KEY) {
    logger.warn('Mem0 API key not configured. Memory features will be disabled.');
    return null;
  }

  httpClient = {
    baseURL: MEM0_API_URL,
    timeout: MEM0_TIMEOUT_MS,
    headers: {
      'Authorization': `Token ${MEM0_API_KEY}`,
      'Content-Type': 'application/json'
    }
  };

  logger.info('Mem0 client initialized', { url: MEM0_API_URL });
  return httpClient;
};

/**
 * Check if Mem0 is available
 * @returns {boolean} True if Mem0 API key is configured
 */
const isAvailable = () => {
  return !!MEM0_API_KEY;
};

/**
 * Test connection to Mem0 API
 * @returns {Promise<boolean>} True if connection is successful
 */
const testConnection = async () => {
  const client = getClient();
  if (!client) {
    return false;
  }

  try {
    // Try to list memories as a health check (with empty user_id returns empty array)
    const response = await makeRequest('GET', '/memories/', { user_id: 'health_check' });
    logger.info('Mem0 connection successful');
    return response !== null;
  } catch (error) {
    logger.error('Mem0 connection failed:', { error: error.message });
    return false;
  }
};

/**
 * Make HTTP request to Mem0 API
 * @param {string} method - HTTP method (GET, POST, etc.)
 * @param {string} path - API path
 * @param {Object} [params] - Query parameters (for GET) or body (for POST)
 * @returns {Promise<Object>} Response data
 */
const makeRequest = async (method, path, params = null) => {
  const client = getClient();
  if (!client) {
    throw new Error('Mem0 client not initialized');
  }

  const url = new URL(path, client.baseURL + '/');
  const options = {
    method,
    headers: client.headers,
    signal: AbortSignal.timeout(client.timeout)
  };

  if (method === 'GET' && params) {
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        url.searchParams.append(key, value);
      }
    });
  } else if (params) {
    options.body = JSON.stringify(params);
  }

  const response = await fetch(url.toString(), options);

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    const message = errorData.message || errorData.detail || response.statusText;
    throw new Error(`Mem0 API error: ${message}`);
  }

  return response.json();
};

/**
 * Normalize user ID (MAC address) to consistent format
 * @param {string} userId - Device MAC address or user identifier
 * @returns {string} Normalized user ID (lowercase)
 */
const normalizeUserId = (userId) => {
  if (!userId) return '';
  // Keep colons/dashes, just lowercase
  return userId.toLowerCase();
};

/**
 * Search memories for a user using semantic search with graph support
 * @param {Object} options - Search options
 * @param {string} options.userId - User identifier (e.g., device MAC)
 * @param {string} [options.query] - Search query (default: comprehensive query)
 * @param {number} [options.limit] - Max memories to return
 * @returns {Promise<Object>} Search results with memories, relations, and entities
 */
const searchMemories = async ({
  userId,
  query = 'What is known about this person, their family, pets, interests, skills, routines, and feelings?',
  limit = MEM0_MEMORY_LIMIT
}) => {
  if (!userId) {
    throw new Error('User ID is required');
  }

  const client = getClient();
  if (!client) {
    return { memories: [], relations: [], entities: [] };
  }

  try {
    const cleanUserId = normalizeUserId(userId);

    const response = await makeRequest('POST', '/memories/search/', {
      query,
      user_id: cleanUserId,
      limit,
      output_format: 'v1.1',
      enable_graph: true
    });

    // Handle both array and v1.1 format responses
    const results = Array.isArray(response) ? response : (response.results || []);
    const topLevelRelations = response.relations || [];

    const memories = [];
    const allEntities = new Map();
    const allRelations = [...topLevelRelations];

    // Extract memories, entities, and relations from each result
    results.forEach(item => {
      if (item.memory) {
        memories.push(item.memory);
      }

      // Extract entities from this memory (graph feature)
      if (item.entities && Array.isArray(item.entities)) {
        item.entities.forEach(e => {
          if (e.name) {
            allEntities.set(e.id || e.name, {
              id: e.id,
              name: e.name,
              type: e.type || 'unknown'
            });
          }
        });
      }

      // Extract relations from this memory (if any)
      if (item.relations && Array.isArray(item.relations)) {
        allRelations.push(...item.relations);
      }
    });

    const entities = Array.from(allEntities.values());

    logger.debug('Mem0 search completed:', {
      userId: cleanUserId,
      memoriesCount: memories.length,
      relationsCount: allRelations.length,
      entitiesCount: entities.length
    });

    return {
      memories,
      relations: allRelations,
      entities
    };
  } catch (error) {
    logger.error('Mem0 search failed:', { error: error.message, userId });
    return { memories: [], relations: [], entities: [] };
  }
};

/**
 * Get all memories for a user (without semantic search)
 * @param {Object} options - Options
 * @param {string} options.userId - User identifier
 * @returns {Promise<Object>} All memories and relations
 */
const getAllMemories = async ({ userId }) => {
  if (!userId) {
    throw new Error('User ID is required');
  }

  const client = getClient();
  if (!client) {
    return { memories: [], relations: [] };
  }

  try {
    const cleanUserId = normalizeUserId(userId);

    const response = await makeRequest('GET', '/memories/', {
      user_id: cleanUserId,
      enable_graph: true
    });

    // API returns array directly, not { results: [...] }
    const memoryArray = Array.isArray(response) ? response : (response.results || []);
    const memories = memoryArray.map(m => m.memory).filter(Boolean);
    const relations = response.relations || [];

    logger.debug('Mem0 getAllMemories:', {
      userId: cleanUserId,
      count: memories.length
    });

    return { memories, relations };
  } catch (error) {
    logger.error('Mem0 getAllMemories failed:', { error: error.message, userId });
    return { memories: [], relations: [] };
  }
};

/**
 * Add conversation messages to memory with graph extraction
 * @param {Object} options - Add options
 * @param {string} options.userId - User identifier
 * @param {Array} options.messages - Array of {role: 'user'|'assistant', content: string}
 * @param {Object} [options.metadata] - Optional metadata (e.g., session_id)
 * @returns {Promise<Object|null>} Result or null on error
 */
const addMemory = async ({ userId, messages, metadata = null }) => {
  if (!userId) {
    throw new Error('User ID is required');
  }

  if (!messages || !Array.isArray(messages) || messages.length === 0) {
    throw new Error('Messages array is required');
  }

  const client = getClient();
  if (!client) {
    return null;
  }

  try {
    const cleanUserId = normalizeUserId(userId);

    // Validate message format
    const validMessages = messages.filter(m =>
      m.role && ['user', 'assistant'].includes(m.role) && m.content
    );

    if (validMessages.length === 0) {
      logger.warn('No valid messages to add to Mem0');
      return null;
    }

    const body = {
      messages: validMessages,
      user_id: cleanUserId,
      enable_graph: true
    };

    if (metadata) {
      body.metadata = metadata;
    }

    const result = await makeRequest('POST', '/memories/', body);

    logger.info('Mem0 addMemory completed:', {
      userId: cleanUserId,
      messagesCount: validMessages.length
    });

    return result;
  } catch (error) {
    logger.error('Mem0 addMemory failed:', { error: error.message, userId });
    return null;
  }
};

/**
 * Add a single fact to memory
 * @param {Object} options - Add options
 * @param {string} options.userId - User identifier
 * @param {string} options.fact - The fact to add
 * @returns {Promise<Object|null>} Result or null on error
 */
const addFact = async ({ userId, fact }) => {
  if (!fact) {
    return null;
  }

  return addMemory({
    userId,
    messages: [{ role: 'user', content: fact }]
  });
};

/**
 * Add conversation transcript to memory (from chat history format)
 * @param {Object} options - Add options
 * @param {string} options.userId - User identifier (e.g., MAC address)
 * @param {Array} options.chatHistory - Array of {chatType: 1|2, content: string}
 * @param {string} [options.sessionId] - Optional session identifier
 * @returns {Promise<Object|null>} Result or null on error
 */
const addConversation = async ({ userId, chatHistory, sessionId = null }) => {
  if (!chatHistory || !Array.isArray(chatHistory) || chatHistory.length === 0) {
    return null;
  }

  // Convert chat history format to Mem0 message format
  // chatType: 1 = user, 2 = agent/assistant
  const messages = chatHistory
    .filter(msg => msg.content && msg.content.trim())
    .map(msg => ({
      role: msg.chatType === 1 ? 'user' : 'assistant',
      content: msg.content.trim()
    }));

  if (messages.length === 0) {
    return null;
  }

  const metadata = sessionId ? { session_id: sessionId } : null;

  return addMemory({
    userId,
    messages,
    metadata
  });
};

/**
 * Delete a specific memory by ID
 * @param {Object} options - Delete options
 * @param {string} options.memoryId - Memory ID to delete
 * @returns {Promise<boolean>} True if successful
 */
const deleteMemory = async ({ memoryId }) => {
  if (!memoryId) {
    throw new Error('Memory ID is required');
  }

  const client = getClient();
  if (!client) {
    return false;
  }

  try {
    await makeRequest('DELETE', `/memories/${memoryId}/`);
    logger.info('Mem0 memory deleted:', { memoryId });
    return true;
  } catch (error) {
    logger.error('Mem0 deleteMemory failed:', { error: error.message, memoryId });
    return false;
  }
};

/**
 * Delete all memories for a user
 * @param {Object} options - Delete options
 * @param {string} options.userId - User identifier
 * @returns {Promise<boolean>} True if successful
 */
const deleteAllMemories = async ({ userId }) => {
  if (!userId) {
    throw new Error('User ID is required');
  }

  const client = getClient();
  if (!client) {
    return false;
  }

  try {
    const cleanUserId = normalizeUserId(userId);

    await makeRequest('DELETE', '/memories/', {
      user_id: cleanUserId
    });

    logger.info('Mem0 all memories deleted:', { userId: cleanUserId });
    return true;
  } catch (error) {
    logger.error('Mem0 deleteAllMemories failed:', { error: error.message, userId });
    return false;
  }
};

/**
 * Get memory by ID
 * @param {Object} options - Get options
 * @param {string} options.memoryId - Memory ID
 * @returns {Promise<Object|null>} Memory object or null
 */
const getMemory = async ({ memoryId }) => {
  if (!memoryId) {
    throw new Error('Memory ID is required');
  }

  const client = getClient();
  if (!client) {
    return null;
  }

  try {
    const result = await makeRequest('GET', `/memories/${memoryId}/`);
    return result;
  } catch (error) {
    if (error.message.includes('404') || error.message.includes('not found')) {
      return null;
    }
    logger.error('Mem0 getMemory failed:', { error: error.message, memoryId });
    throw error;
  }
};

/**
 * Update a memory
 * @param {Object} options - Update options
 * @param {string} options.memoryId - Memory ID
 * @param {string} options.text - New memory text
 * @returns {Promise<Object|null>} Updated memory or null
 */
const updateMemory = async ({ memoryId, text }) => {
  if (!memoryId) {
    throw new Error('Memory ID is required');
  }

  if (!text) {
    throw new Error('Text is required');
  }

  const client = getClient();
  if (!client) {
    return null;
  }

  try {
    const result = await makeRequest('PUT', `/memories/${memoryId}/`, {
      text
    });

    logger.info('Mem0 memory updated:', { memoryId });
    return result;
  } catch (error) {
    logger.error('Mem0 updateMemory failed:', { error: error.message, memoryId });
    return null;
  }
};

/**
 * Get memory history (previous versions)
 * @param {Object} options - Options
 * @param {string} options.memoryId - Memory ID
 * @returns {Promise<Array>} History entries
 */
const getMemoryHistory = async ({ memoryId }) => {
  if (!memoryId) {
    throw new Error('Memory ID is required');
  }

  const client = getClient();
  if (!client) {
    return [];
  }

  try {
    const result = await makeRequest('GET', `/memories/${memoryId}/history/`);
    return Array.isArray(result) ? result : (result.results || []);
  } catch (error) {
    logger.error('Mem0 getMemoryHistory failed:', { error: error.message, memoryId });
    return [];
  }
};

/**
 * Format memories for prompt injection
 * @param {Object} memoryData - Memory data from searchMemories
 * @returns {string} Formatted string for prompt
 */
const formatForPrompt = (memoryData) => {
  if (!memoryData || !memoryData.memories || memoryData.memories.length === 0) {
    return '';
  }

  const lines = ['## What I know about you:', ''];

  // Add memories
  memoryData.memories.forEach(memory => {
    lines.push(`- ${memory}`);
  });

  // Add relations if present
  if (memoryData.relations && memoryData.relations.length > 0) {
    lines.push('');
    lines.push('### Relationships:');
    memoryData.relations.forEach(rel => {
      if (rel.source && rel.relation && rel.target) {
        lines.push(`- ${rel.source} ${rel.relation} ${rel.target}`);
      }
    });
  }

  return lines.join('\n');
};

/**
 * Reset HTTP client (for testing)
 */
const resetClient = () => {
  httpClient = null;
};

module.exports = {
  // Client management
  getClient,
  isAvailable,
  testConnection,
  resetClient,

  // Memory operations
  searchMemories,
  getAllMemories,
  addMemory,
  addFact,
  addConversation,
  deleteMemory,
  deleteAllMemories,
  getMemory,
  updateMemory,
  getMemoryHistory,

  // Utilities
  normalizeUserId,
  formatForPrompt,

  // Constants
  MEM0_MEMORY_LIMIT,
  MEM0_TIMEOUT_MS
};
